Partial sink pipelines should do something sensible with multiple calls

isiahmeadows opened the pull request:
The extra work is necessary to avoid keeping garbage lying around. I did make an effort to avoid an extra array allocation for the common case of smaller partial pipelines (5 or less in the chain), though.

isiahmeadows commented:
Still working on fixing this. Not ready to merge...

isiahmeadows commented:
Ready.

dominictarr commented:
what problem does this solve?

I don't think this is necessary because you should never use a particular stream in two pipelines,
you should create new streams...

dominictarr commented:
oh, do you mean garbage as in things that will be GC'd, so this saves creating an array and then throwing it out? is it possible to make a script/benchmark that shows how this is a problem?

isiahmeadows commented:
@dominictarr This is solving two problems:

1. You can reuse a `pull()` sink. It might not always be a good idea, but it could be a thing.
2. The callback is more easily collectable.

The `switch` statement was actually just avoiding an additional array allocation for shorter things. If you want a benchmark, it's trivial to write, but I pulled it since it's unlikely to be a problem in practice - you'll have to do `reader(transform(readable))` by that point.

dominictarr commented:
reusing a sink chain would definitely be at your own risk and not advised in general.
pull streams often have internal state (i.e. `highwatermark`, or `take`,
and these would have unintended behavior when reused.

isiahmeadows commented:
The `read` callback may never get cleared otherwise until after the current tick ends, and it could cause a slow memory leak if you're not careful.

dominictarr commented:
regards the GC thing though, it would be interesting to see _how much_ difference it does make, so a benchmark is appreciated. we have one here https://github.com/pull-stream/pull-stream/blob/master/benchmarks/pull.js from the last time someone made an optimization pull-request.

isiahmeadows commented:
I benched three different methods for that:

1. The original `master`, without this patch.
2. Replacing the body of the returned function with `push.apply(null, [read].concat(args))`.
3. The `switch` statement I'm currently using in it.

Performance-wise, here's what I've found:

- `pull3` is about equal across all of them. I suspect the code path isn't hit.
- `pull_compose` is about 5-10% faster with the `switch`, but 5-10% slower with the `concat`. I'm pretty sure that returned function is hit in the second `pull` call of [this line](https://github.com/pull-stream/pull-stream/blob/master/benchmarks/pull.js#L35).
- `pull_chain` is roughly equal between the original and `concat`, but the `switch` version is slightly faster (not enough to mean much in practice).

Here's the results of 3 consecutive runs of each case:

Current `master`:

```
pull3*100000: 4073.384ms
pull_compose*100000: 4474.739ms
pull_chain*100000: 4008.915ms
pull3*100000: 4017.678ms
pull_compose*100000: 4432.934ms
pull_chain*100000: 3945.300ms
pull3*100000: 3983.170ms
pull_compose*100000: 4475.936ms
pull_chain*100000: 3952.188ms
```

This patch, using `concat`:

```
pull3*100000: 3902.641ms
pull_compose*100000: 4683.064ms
pull_chain*100000: 3979.576ms
pull3*100000: 4065.228ms
pull_compose*100000: 4761.607ms
pull_chain*100000: 4014.491ms
pull3*100000: 4187.816ms
pull_compose*100000: 4693.648ms
pull_chain*100000: 4020.804ms
```

This patch, using a `switch` statement:

```
pull3*100000: 4005.613ms
pull_compose*100000: 4108.665ms
pull_chain*100000: 3855.271ms
pull3*100000: 3972.441ms
pull_compose*100000: 4259.308ms
pull_chain*100000: 3931.466ms
pull3*100000: 3951.960ms
pull_compose*100000: 4152.017ms
pull_chain*100000: 3872.062ms
```

Here are the times, reformatted as a table:

Benchmark    | `master`      | This with `concat` | This with `switch`
-------------|---------------|--------------------|--------------------
pull3        | 1. 4073.384ms | 1. 3902.641ms      | 1. 4005.613ms
             | 2. 4017.678ms | 2. 4065.228ms      | 2. 3972.441ms
             | 3. 3983.170ms | 3. 4187.816ms      | 3. 3951.960ms
pull_compose | 1. 4474.739ms | 1. 4683.064ms      | 1. 4108.665ms
             | 2. 4432.934ms | 2. 4761.607ms      | 2. 4259.308ms
             | 3. 4475.936ms | 3. 4693.648ms      | 3. 4152.017ms
pull_chain   | 1. 4008.915ms | 1. 3979.576ms      | 1. 3855.271ms
             | 2. 3945.300ms | 2. 4014.491ms      | 2. 3931.466ms
             | 3. 3952.188ms | 3. 4020.804ms      | 3. 3872.062ms

dominictarr commented:
looks good, but I think this can use https://github.com/pull-stream/pull-stream/pull/58/files#diff-2e88b737a946cc8fb8aacd25ff7ffe5aR15 `unshift` as in master instead of `concat`.

I just want to be completely clear that calling reusing the sink chain is not supported.
I'd sooner merge a thing that threw if called you `sink(read)` more than once.

isiahmeadows commented:
@dominictarr I added that check and asserted it throws instead. I also changed the last branch (slow path) of the `switch` + the test to use `unshift` for consistency. Oh, and I also managed to speed it up further by using a local alias to the arguments. Here's the latest benchmark numbers (`node benchmark/pull.js` 3 times):

```
pull3*100000: 3891.233ms
pull_compose*100000: 4133.924ms
pull_chain*100000: 3867.556ms
pull3*100000: 3890.908ms
pull_compose*100000: 4076.031ms
pull_chain*100000: 3866.232ms
pull3*100000: 3818.921ms
pull_compose*100000: 4040.969ms
pull_chain*100000: 3815.148ms
```

In a table, compared to `master`:

Benchmark    | `master`      | Current
-------------|---------------|---------------
pull3        | 1. 4073.384ms | 1. 3891.233ms
             | 2. 4017.678ms | 2. 3890.908ms
             | 3. 3983.170ms | 3. 3818.921ms
pull_compose | 1. 4474.739ms | 1. 4133.924ms
             | 2. 4432.934ms | 2. 4076.031ms
             | 3. 4475.936ms | 3. 4040.969ms
pull_chain   | 1. 4008.915ms | 1. 3867.556ms
             | 2. 3945.300ms | 2. 3866.232ms
             | 3. 3952.188ms | 3. 3815.148ms

dominictarr commented:
sorry i let this slip, merged into 3.4.4
